jscomp/runtime/js.mli

```diff
@@ -0,0 +1,208 @@
+
+module Internal : sig
+  external opaqueFullApply : 'a -> 'a = "#full_apply"
+  external opaque : 'a -> 'a = "%opaque"
+
+  external run : 'a Fn.arity0 -> 'a = "#run"
+  [@@ocaml.warning "-unboxable-type-in-prim-decl"]
+end
+
+type +'a null = 'a Js_null.t
+(** nullable, value of this type can be either [null] or ['a]
+    this type is the same as type [t] in {!Null}
+*)
+
+type +'a undefined = 'a Js_undefined.t
+(** value of this type can be either [undefined] or ['a]
+    this type is the same as type [t] in {!Undefined}  *)
+
+type +'a nullable = 'a Js_null_undefined.t
+(** value of this type can be [undefined], [null] or ['a]
+    this type is the same as type [t] n {!Null_undefined} *)
+
+type +'a null_undefined = 'a nullable
+
+external toOption : 'a nullable -> 'a option = "#nullable_to_opt"
+external undefinedToOption : 'a undefined -> 'a option = "#undefined_to_opt"
+external nullToOption : 'a null -> 'a option = "#null_to_opt"
+external isNullable : 'a nullable -> bool = "#is_nullable"
+
+external testAny : 'a -> bool = "#is_nullable"
+(** The same as {!test} except that it is more permissive on the types of input *)
+
+external null : 'a null = "#null"
+(** The same as [empty] in {!Js.Null} will be compiled as [null]*)
+
+external undefined : 'a undefined = "#undefined"
+(** The same as  [empty] {!Js.Undefined} will be compiled as [undefined]*)
+
+external typeof : 'a -> string = "#typeof"
+(** [typeof x] will be compiled as [typeof x] in JS
+    Please consider functions in {!Types} for a type safe way of reflection
+*)
+
+external log : 'a -> unit = "log"
+[@@mel.scope "console"]
+(** A convenience function to log everything *)
+
+external log2 : 'a -> 'b -> unit = "log" [@@mel.scope "console"]
+external log3 : 'a -> 'b -> 'c -> unit = "log" [@@mel.scope "console"]
+external log4 : 'a -> 'b -> 'c -> 'd -> unit = "log" [@@mel.scope "console"]
+
+external logMany : 'a array -> unit = "log"
+[@@mel.scope "console"] [@@mel.splice]
+(** A convenience function to log more than 4 arguments *)
+
+external eqNull : 'a -> 'a null -> bool = "%bs_equal_null"
+external eqUndefined : 'a -> 'a undefined -> bool = "%bs_equal_undefined"
+external eqNullable : 'a -> 'a nullable -> bool = "%bs_equal_nullable"
+
+(** {4 operators }*)
+
+external unsafe_lt : 'a -> 'a -> bool = "#unsafe_lt"
+(** [unsafe_lt a b] will be compiled as [a < b].
+    It is marked as unsafe, since it is impossible
+    to give a proper semantics for comparision which applies to any type
+*)
+
+external unsafe_le : 'a -> 'a -> bool = "#unsafe_le"
+(**  [unsafe_le a b] will be compiled as [a <= b].
+     See also {!unsafe_lt}
+*)
+
+external unsafe_gt : 'a -> 'a -> bool = "#unsafe_gt"
+(**  [unsafe_gt a b] will be compiled as [a > b].
+     See also {!unsafe_lt}
+*)
+
+external unsafe_ge : 'a -> 'a -> bool = "#unsafe_ge"
+(**  [unsafe_ge a b] will be compiled as [a >= b].
+     See also {!unsafe_lt}
+*)
+
+external unsafe_downgrade : 'a t -> 'a = "#unsafe_downgrade"
```

Comment:
This is prob the most controversial part of the PR, but before it was already accessible to consumers through `Js__.Js_OO.unsafe_downgrade`, so nothing has changed in terms of visibility.

Reply:
Ah, and we need it in `js.ml` because it needs `Js.t`. That's fine.